### ISSUE TITLE
feature: support Encrypted ClientHello

### ISSUE THREAD

**Issue by diracdeltas** [brave/brave-browser#1851]
per slack discussion from a while ago, we should support encrypted SNI on all platforms: https://blog.mozilla.org/security/2018/10/18/encrypted-sni-comes-to-firefox-nightly/

**Comment by DavidRBailey** [comment]
Encrypted SNI is getting attention. https://medium.com/asecuritysite-when-bob-met-alice/your-coffee-shop-wi-fi-can-see-what-you-are-connecting-to-8c2a441be6ae

**Comment by HLFH** [comment]
Related to https://github.com/brave/browser-laptop/issues/12235
Please follow: https://bugs.chromium.org/p/chromium/issues/detail?id=908132
And the IETF TLS working group Internet Draft: https://github.com/tlswg/draft-ietf-tls-esni

**Comment by rogersachan** [comment]
https://www.cloudflare.com/ssl/encrypted-sni/ seems to emphasize having support for this alongside DNS over HTTPS (#1864)

**Comment by jadchaar** [comment]
Chromium 78 shipped this week with an [optional flag to enable DNS over HTTPS](https://blog.chromium.org/2019/09/experimenting-with-same-provider-dns.html), so I think it would be great to get encrypted SNI soon as well.

**Comment by riastradh-brave** [comment]
We are totally on board with adopting encrypted SNI as soon as we can.

According to https://bugs.chromium.org/p/chromium/issues/detail?id=908132, the state of ESNI in Chromium seems to be that Google is waiting for the IETF TLS working group to finish the draft https://datatracker.ietf.org/doc/draft-ietf-tls-esni/ before they implement it.  (The last update to the draft was yesterday, so it evidently hasn't settled yet.)

Since it is likely to involve some intimate surgery in the TLS stack in BoringSSL, we probably won't do this ourselves if Google is likely to do it soon anyway.  Probably the best way to get things going is to let the IETF TLS WG finish the specification as soon as possible.

**Comment by uwe22** [comment]
What is the status of privacy regarding ESNI? 
Does the Brave Browser support encrypted SNI now?
For people living in countries filtering content by SNI this is crucial.

![brave esni](https://user-images.githubusercontent.com/11908297/76136196-5f527880-6072-11ea-9d7b-89b96eb6a1bf.jpg)

**Comment by csoko** [comment]
> What is the status of privacy regarding ESNI?
> Does the Brave Browser support encrypted SNI now?
> For people living in countries filtering content by SNI this is crucial.
> 
> ![brave esni](https://user-images.githubusercontent.com/11908297/76136196-5f527880-6072-11ea-9d7b-89b96eb6a1bf.jpg)

Use firefox for the time being....
![image](https://user-images.githubusercontent.com/22408357/77784128-41e05100-7017-11ea-87d3-acce1a6f48d5.png)

**Comment by AnkS4** [comment]
Is there any progress on ESNI?

**Comment by xenstar** [comment]
Yeah me too, also want to know. Is there any progress? Brave is great, but still using firefox because I can get full ESNI support.

**Comment by FatBirdie** [comment]
Yeah, status update pllzzzz

**Comment by MrClayPole** [comment]
Any update on this feature?

**Comment by octavian-negru** [comment]
> Any update on this feature?

+1

**Comment by FatBirdie** [comment]
I would also like to know if there are any updates for this

**Comment by xenstar** [comment]
So many people are waiting for this, hopefully brave will add it soon

**Comment by MihaiVoinea** [comment]
Any updates on this?

**Comment by Genkilabs** [comment]
+1

**Comment by evangallo** [comment]
+1, it seems like the specification is taking forever - would be great to see support for this sooner rather than later.

**Comment by trojek** [comment]
+1

**Comment by alexandreszczepaniak** [comment]
+1

**Comment by uwe22** [comment]
+1

**Comment by keerah** [comment]
this is the only feature that holds me off using Brave as my everyday browser

**Comment by vivektaparia** [comment]
This is needed in Brave, hope this get to Brave soon.

**Comment by FireMasterK** [comment]
Is this gonna be fixed by Brave or Chromium? This literally breaks censorship in many countries and at the same time, hides what you're doing from your ISP.

**Comment by bsclifton** [comment]
@FireMasterK this would be fixed upstream in Chromium

It appears some good work has been done on the Chromium side since then. The issues tracking this work are:
- https://bugs.chromium.org/p/chromium/issues/detail?id=1091403
- https://bugs.chromium.org/p/boringssl/issues/detail?id=275

You can login with a Google account and star those and you'll get email notifications on update

**Comment by ShrirajHegde** [comment]
So this isn't coming to **Linux** until chromium gets it? says disabled on the platform (literally all other platforms supported by chromium are supported)

**Comment by Snorkeldink** [comment]
+1

**Comment by kaiyang-optimize** [comment]
+1

**Comment by jawadalkassim** [comment]
ESNI will be unsupported soon , they will replace it with ECH
https://blog.cloudflare.com/encrypted-client-hello/
DOH without ESNI or Ech is useless

**Comment by Ezols** [comment]
Its been already quiet a time, how is it going to implement SNI encryption in this security browser?

**Comment by blastagator** [comment]
> Its been already quiet a time, how is it going to implement SNI encryption in this security browser?
It is very unlikely this will happen now that ESNI is deprecated.

**Comment by FireMasterK** [comment]
Is there an issue to track ECH tech on brave?

**Comment by v1shva** [comment]
> > Its been already quiet a time, how is it going to implement SNI encryption in this security browser?
> 
> It is very unlikely this will happen now that ESNI is deprecated.

Can you post a link to the source? Why have they deprecated this? ESNI and DNS over HTTPS should be the default for brave.

**Comment by blastagator** [comment]
> > > Its been already quiet a time, how is it going to implement SNI encryption in this security browser?
> > 
> > 
> > It is very unlikely this will happen now that ESNI is deprecated.
> 
> Can you post a link to the source? Why have they deprecated this? ESNI and DNS over HTTPS should be the default for brave.

Sorry, should have clarified.  It looks like ESNI was dropped in favor of something called ECH.  A good blog post about it is here: https://blog.cloudflare.com/encrypted-client-hello/

I wouldn't expect it to be ready for production for awhile.  These things are rather complicated and require formalized standards across the industry.

**Comment by diracdeltas** [comment]
FYI DoH can now be enabled in brave via brave://settings/security

here's the chromium tracking issue for ECH: https://bugs.chromium.org/p/chromium/issues/detail?id=1091403

it may not be feasible for us to ship it before Chromium does if it requires a lot of changes in the Chromium TLS code and is blocked on changes to BoringSSL (https://bugs.chromium.org/p/boringssl/issues/detail?id=275)

**Comment by Lim4** [comment]
Something new about ClientHello?

**Comment by AaronStiff** [comment]
Any updates? :)

**Comment by terrytw** [comment]
I think it is drawing close, since Chromium has included it in 104 version.

You can now manually set the flag of ECH.

**Comment by gaby64** [comment]
> I think it is drawing close, since Chromium has included it in 104 version.
> 
> You can now manually set the flag of ECH.

how? whats the flag?

**Comment by xenstar** [comment]
> > I think it is drawing close, since Chromium has included it in 104 version.
> > You can now manually set the flag of ECH.
> 
> how? whats the flag?

brave://flags/ I think he meant here. Now I have 104 version, but didn't see anything with Client Hello

**Comment by terrytw** [comment]
> > > I think it is drawing close, since Chromium has included it in 104 version.
> > > You can now manually set the flag of ECH.
> > 
> > 
> > how? whats the flag?
> 
> brave://flags/ I think he meant here. Now I have 104 version, but didn't see anything with Client Hello

Download Chromium https://github.com/chromium/chromium/tags
Also after enabling it, I went to Cloudflare's check page (https://www.cloudflare.com/zh-cn/ssl/encrypted-sni/#results), it still shows that ESNI not enabled. I guess it is still early days.

You can check the development here:
https://bugs.chromium.org/p/chromium/issues/detail?id=1091403

**Comment by ShivanKaul** [comment]
Chromium 105 has the `#encrypted-client-hello` feature flag, Brave will be moving to that shortly.

**Comment by ShivanKaul** [comment]
Encrypted ClientHello is available on Brave Nightly now, but requires some setup:
1. Make sure you have DoH turned on in brave://settings/security with a provider chosen
2. In brave://flags, search for "Encrypted ClientHello" and "Support for HTTPS records in DNS" and enable them.
3. Restart the browser

ECH should work now. You can try it out by going to https://defo.ie/ech-check.php or any other ECH checker.

**Comment by xenstar** [comment]
> Encrypted ClientHello is available on Brave Nightly now, but requires some setup:
> 
> 1. Make sure you have DoH turned on in brave://settings/security with a provider chosen
> 2. In brave://flags, search for "Encrypted ClientHello" and "Support for HTTPS records in DNS" and enable them.
> 3. Restart the browser
> 
> ECH should work now. You can try it out by going to https://defo.ie/ech-check.php or any other ECH checker.

Very happy to hear that. Some of us are using AdguardHome and enabled the nextdns DoH on the network level. In that case, do we still need to enable DoH in browser?

**Comment by xenstar** [comment]
![image](https://user-images.githubusercontent.com/55502651/187081611-d92f7352-5b46-4132-8de9-b59d56c5d861.png)
1) DNS is set to NEXTDNS DoH

![image](https://user-images.githubusercontent.com/55502651/187081702-a681246e-8c2d-4337-98cf-71d4fd9b0c74.png)
2) brave://flags (already enabled)

3) Restart Browser. Done!

ECH still not working

![image](https://user-images.githubusercontent.com/55502651/187081712-1fc64aad-a91c-452e-bc44-8c86acf80b1d.png)

![Uploading image.png…]()

![Uploading image.png…]()

**Comment by xenstar** [comment]
I have downloaded the latest nightly version. Here is the feedback after 9 days. 

**This is my settings:** 

![image](https://user-images.githubusercontent.com/55502651/188257872-a907b62d-9a4f-461c-b3ad-aaede2ad671c.png)

![image](https://user-images.githubusercontent.com/55502651/188257883-d349c7ab-c7db-4ae6-9816-0f03491a53db.png)

**But ECH is not working**

![image](https://user-images.githubusercontent.com/55502651/188257894-37063e97-ddb8-4a09-bb71-c86cd4d55a91.png)

![image](https://user-images.githubusercontent.com/55502651/188257898-80db50c2-e4e3-4c4a-807f-8eed879c7a55.png)

![image](https://user-images.githubusercontent.com/55502651/188257906-0a389f97-a7bb-47c6-b3ca-ecaf7d5e6542.png)

**Comment by ShivanKaul** [comment]
The Cloudflare test page, while incorrect for ECH (at least last time I checked) is indicating there's something with your DoH connection. Could you try with a different provider? Or generally troubleshoot DoH first

**Comment by sergey-alekseev** [comment]
@ShivanKaul why do all test results show that Encrypted (Secure) SNI is not working? Seems like the feature doesn't really work anywhere (I tried Brave, Chrome and Firefox).

**Comment by 0x1a8510f2** [comment]
@sergey-alekseev because these browsers no longer use ESNI (encrypted SNI), but rather, ECH (encrypted ClientHello) which is meant to be more secure. Tests which correctly use ECH should work.

**Comment by sergey-alekseev** [comment]
@0x1a8510f2 https://tls-ech.dev/ says "You are not using ECH. :(" same as on the screenshot above. 

>Tests which correctly use ECH should work.

Where would you suggest to test?

**Comment by gregtwallace** [comment]
Did you enable "use secure dns" and the two experimental flags?  My PC defaults to using encrypted DNS but ECH still didn't work until I manually configured it in the brower (which is obnoxious).

**Comment by sergey-alekseev** [comment]
Yes  
<img width="665" alt="Screenshot 2022-11-11 at 20 18 32" src="https://user-images.githubusercontent.com/2371639/201394476-2c3bdebc-57ef-4d6b-ae37-b67657ae9780.png">
<img width="742" alt="Screenshot 2022-11-11 at 20 19 01" src="https://user-images.githubusercontent.com/2371639/201394554-c9962f21-84c0-4884-898c-c16605d7d2bd.png">
<img width="700" alt="Screenshot 2022-11-11 at 20 19 17" src="https://user-images.githubusercontent.com/2371639/201394601-a0898118-79b3-4a80-993f-57830eb2abc8.png">

Basically like [above](#issuecomment-1236055639)
@gregtwallace

**Comment by sergey-alekseev** [comment]
Finally, I figured this out with your help – the problem is caused by third-party apps which do HTTPS filtering. When I disable all HTTPS filtering then everything works as expected. Hope this will help someone who has the same problem.

**Comment by sergey-alekseev** [comment]
Thanks @simpvn. The problem in my case is caused by third-party apps which do HTTPS filtering. You can see right [above](#issuecomment-1311973701) in case you missed my previous comment.

**Comment by xenstar** [comment]
> When I disable all HTTPS filtering, then everything works as expected.

Hi Sergey, I wonder how you figured out which 3rd party applications are filtering HTTPS traffic and how you turned it off? I am using windows 11. if you could give me any information would be great. I have been waiting for many days to enable ECH.

**Comment by sergey-alekseev** [comment]
@xenstar in my case it was two apps – an antivirus and an ad blocker. After I turn them off ECH works. I reached out to developers of those apps mentioning this problem.

**Comment by 1ulzsec** [comment]
I just found the cause for this problem.
I'm using a vpn, and the problem was solved when I tested the ECH again after disconnecting the vpn :)

![image](https://user-images.githubusercontent.com/114208850/212578511-e9762a7f-3ef0-4e61-a78a-6d63aaf90725.png)

**Comment by ValZapod** [comment]
>I'm using a vpn, and the problem was solved when I tested the ECH again after disconnecting the vpn :)

True.

**Comment by ValZapod** [comment]
>I'm using a vpn, and the problem was solved when I tested the ECH again after disconnecting the vpn :)

Same here. #dns-https-svcb does not exist anymore too.

**Comment by ShivanKaul** [comment]
Yep, there's been an update. Now you need the following flags:
1. brave://flags/#use-dns-https-svcb-alpn
2. brave://flags/#encrypted-client-hello

DoH is not mandatory any more.

**Comment by ValZapod** [comment]
>DoH is not mandatory any more.

It was not mandatery before either. You could select plaintext DNS.

**Comment by asheroto** [comment]
FYI looks like Cloudflare was one of the authors of the [IETF doc](https://datatracker.ietf.org/doc/draft-ietf-tls-esni/) which was renamed .

> * brave://flags/#use-dns-https-svcb-alpn
> * brave://flags/#encrypted-client-hello

As of June 2023, on desktop, this worked for me. Using DoH.

On the Android version, however, [this test](https://www.cloudflare.com/ssl/encrypted-sni/#results) shows that it is **NOT** enabled. Can anyone else confirm? Enabled the flags, relaunched, tried it with and without DoH.

**Comment by kdenhartog** [comment]
Chromium is currently testing this at the moment so we're likely to see this incoming soon. They're currently testing with 1% of chrome instances (50%/50% on pre-release channels) and the expected end of the trial is the 09/11/23.

**Comment by SrGeneroso** [comment]
I've enable the flag in [brave flags](chrome://flags/?search=Encrypted%20ClientHello#encrypted-client-hello)
[ brave://flags/ in ](chrome://flags/?search=Encrypted%20ClientHello#encrypted-client-hello)
How could I know if the page is using it? I guess not all webs support it and I wonder if there is a easy way to test if it was used in the current web I'm looking at.

**Comment by kreaninw** [comment]
> I've enable the flag in brave flags brave://flags/ in How could I know if the page is using it? I guess not all webs support it and I wonder if there is a easy way to test if it was used in the current web I'm looking at.

I enabled `#encrypted-client-hello` in `brave://flags`. It's working correctly. You should be able to check whether it's working at: https://www.cloudflare.com/ssl/encrypted-sni/

If it's working, the Secure SNI AKA Encrypted Client Hello (ECH) should have the right mark ticked.

![TinySnap-2023-10-14-17 45 27](https://github.com/brave/brave-browser/assets/1403194/97ad477f-676c-49c2-9f05-bb62f25a76a8)

**Comment by asheroto** [comment]
For some reason my desktop version isn't working anymore on Cloudflare...

![image](https://github.com/brave/brave-browser/assets/49938263/9473a5c5-24c3-467e-8efe-6290d4af7836)

![image](https://github.com/brave/brave-browser/assets/49938263/0ec349a9-0503-4a1e-b154-b5f7de3f69a8)

![image](https://github.com/brave/brave-browser/assets/49938263/125b6a43-e954-4521-a355-28309ed9b33e)

oddly enough, tls-ech.dev reports it's fine...

![image](https://github.com/brave/brave-browser/assets/49938263/2b993898-d1c3-4f85-9d7a-5dde2ee44172)

These are the same thing, right?

![image](https://github.com/brave/brave-browser/assets/49938263/49c1b463-898e-4135-b0ad-1259e1d64492)

Latest version

**Comment by 097115** [comment]
@asheroto, by any chance, you don't have something like Warp turned on?

**Comment by asheroto** [comment]
> @asheroto, by any chance, you don't have something like Warp turned on?

No Warp, no VPN, no HTTPS filtering or site filtering. Thinking maybe something quicky with the Cloudflare test since tls-ech.dev works on both phone & desktop now. Also if I go to the following URL, I see `sni=encrypted` whereas on Edge it says plaintext.

https://crypto.cloudflare.com/cdn-cgi/trace

**Comment by kreaninw** [comment]
@asheroto Your ECH works correctly even if the Cloudflare test says otherwise.

However, from my understanding, ECH needs to be enabled on both sides, i.e. on the client side(browser) and the server side(website),  it won't work otherwise. ECH would benefit VPN users the most, logging or not, we don't have to trust the VPN providers all that much since they won't know the websites we have gone through, see https://blog.mozilla.org/en/products/firefox/encrypted-hello/. That could be the reason why Cloudflare Warp doesn't support ECH. 

I am using Warp and won't trade my encrypted tunnel with ECH.   Warp hides me from my ISP(ECH would be better since I don't have to put my trust in anyone), and also hides my IP from the other end(the DoH, DoT, and ECH combo doesn't).

**Comment by asheroto** [comment]
Thanks for the info.

Looks like we're good to go. I'm relying on [tls-ech.dev](https://tls-ech.dev/) and the [Cloudflare trace page](https://crypto.cloudflare.com/cdn-cgi/trace) to confirm ECH support. Phone and desktop both show enabled now.

I actually use Cloudflare for my sites. For anyone interested that also uses Cloudflare, there's an option in the dashboard where you can [enable ECH for server-side](https://developers.cloudflare.com/ssl/edge-certificates/ech/#enable-ech), however it is [currently disabled](https://community.cloudflare.com/t/early-hints-and-encrypted-client-hello-ech-are-currently-disabled-globally/567730) due to some bugs.

**Comment by mich2k** [comment]
> @asheroto Your ECH works correctly even if the Cloudflare test says otherwise.
> 
> However, from my understanding, ECH needs to be enabled on both sides, i.e. on the client side(browser) and the server side(website), it won't work otherwise. ECH would benefit VPN users the most, logging or not, we don't have to trust the VPN providers all that much since they won't know the websites we have gone through, see https://blog.mozilla.org/en/products/firefox/encrypted-hello/. That could be the reason why Cloudflare Warp doesn't support ECH.
> 
> I am using Warp and won't trade my encrypted tunnel with ECH. Warp hides me from my ISP(ECH would be better since I don't have to put my trust in anyone), and also hides my IP from the other end(the DoH, DoT, and ECH combo doesn't).

Personally I am still going with a self-hosted vpn but was wondering:

ECH + vpn: in this way the ECH must be supported from only the vpn provider (and we would get ECH also on non enabled websites) or still each website we visit has to support ECH (in this way vpn provides will still be able to see traffic on every server non ech-enabled)?

**Comment by asheroto** [comment]
The VPN will work regardless of ECH, as VPN encapsulates the traffic it carriers. But as far as DPI (deep packet inspection) and monitoring which websites are visited, that could be affected by the use of ECH but that's up to the technology you use to monitor sites. And in any case, it won't prevent the traffic from flowing, but it may affect whether you can see what sites people visited since ECH encrypted the SNI (server name indicator).

For example, Ubiquiti uses "proprietary technology" to determine which websites were visited when using their routers. It's been suspected that they may intercept the SNI so that it can report to network administrators which sites were accessed. If ECH is enabled and used, it may limit Ubiquiti's detection capabilities, however it could use other means like recording which IP addresses are commonly associated with specific domain names.

Cloudflare Warp provides a way to filter traffic by determining which websites are being visited, then making a determination on if that website is allowed. If may not support ECH right now because that appears to be the main way they detect the sites, but maybe at a later date it could be supported through other means.

With SonicWALL SSL-DPI, you actually install a client certificate on each device on the network, then the SonicWALL itself creates its own SSL cert with each website, therefore able to decrypt the traffic. Cloudflare Warp may end up doing something like that, too.

This is a bit off topic though. 😊

@diracdeltas This issue may be able to be closed now as it appears that the support is working on all platforms.

**Comment by Nurmagoz** [comment]
I dont think this is the case anymore, ECH/ESNI doesnt work with brave browser.

Websites to test:

- https://www.cloudflare.com/ssl/encrypted-sni
- https://tls-ech.dev/
- https://test.defo.ie/iframe_tests.html

All of them shows that brave is not encrypting anything (whereas i tested that on Firefox, it all show success readings). 

Brave version: v1.78.97 (May 7, 2025), using flatpak.

**Comment by ShivanKaul** [comment]
Can you check you don't have a VPN enabled?

**Comment by Reprevise** [comment]
I'm not on a VPN— here are my results:

| Browser | [Cloudflare 1](https://www.cloudflare.com/ssl/encrypted-sni/) | [Cloudflare 2](https://crypto.cloudflare.com/cdn-cgi/trace) | [tls-ech.dev](https://tls-ech.dev/)
| --- | :---: | :---: | :---: |
| Brave | ❌ | ❌ | ✔ |
| Firefox | ✔ | ❌ | ✔ |

**Comment by Nurmagoz** [comment]
> Can you check you don't have a VPN enabled?

Yeah i dont, im pretty positive you gonna have the same readings specially when testing against [cloudflare esni checker](https://www.cloudflare.com/ssl/encrypted-sni/).
